core/src/peer_id.rs

```diff
@@ -77,44 +102,86 @@ impl PeerId {
     pub fn from_multihash(multihash: Multihash) -> Result<PeerId, Multihash> {
         match Code::try_from(multihash.code()) {
-            Ok(Code::Sha2_256) => Ok(PeerId { multihash }),
-            Ok(Code::Identity) if multihash.digest().len() <= MAX_INLINE_KEY_LENGTH => {
-                Ok(PeerId { multihash })
-            }
+            Ok(Code::Sha2_256) => Ok(PeerId {
+                cid: Cid::new_v1(LIBP2P_KEY_CODEC, multihash),
+            }),
+            Ok(Code::Identity) if multihash.digest().len() <= MAX_INLINE_KEY_LENGTH => Ok(PeerId {
+                cid: Cid::new_v1(LIBP2P_KEY_CODEC, multihash),
+            }),
             _ => Err(multihash),
         }
     }
 
+    /// Tries to turn a `Cid` into a `PeerId`.
+    ///
+    /// If the CID's multihash does not fit the specifications:
+    ///
+    /// * use a valid hashing algorithm for peer IDs, or
+    /// * the value does not satisfy the constraints for a hashed peerID,
+    ///
+    /// or, if the CID does not fit the specifications:
+    ///
+    /// * is a CIDv0 with a multihash that fits the specifications
+    /// for multihash, or
+    /// * is a CIDv1 with the proper codec (LIBP2P_KEY_CODEC),
+    ///
+    /// then the CID will be returned as an `Err`.
+    pub fn from_cid(cid: Cid) -> Result<PeerId, Cid> {
+        let multihash = *cid.hash();
+
+        match cid.version() {
+            Version::V0 => PeerId::from_multihash(multihash).map_err(|_| cid),
+            Version::V1 => {
+                if cid.codec() == LIBP2P_KEY_CODEC {
+                    PeerId::from_multihash(multihash).map_err(|_| cid)
+                } else {
+                    Err(cid)
+                }
+            }
+        }
+    }
+
     /// Generates a random peer ID from a cryptographically secure PRNG.
     ///
     /// This is useful for randomly walking on a DHT, or for testing purposes.
     pub fn random() -> PeerId {
         let peer_id = rand::thread_rng().gen::<[u8; 32]>();
+        let multihash = Multihash::wrap(Code::Identity.into(), &peer_id)
+            .expect("The digest size is never too large");
         PeerId {
-            multihash: Multihash::wrap(Code::Identity.into(), &peer_id)
-                .expect("The digest size is never too large"),
+            cid: Cid::new_v1(LIBP2P_KEY_CODEC, multihash),
         }
     }
 
-    /// Returns a raw bytes representation of this `PeerId`.
+    /// Returns a raw bytes representation of the contents of the `PeerId`.
     pub fn to_bytes(&self) -> Vec<u8> {
-        self.multihash.to_bytes()
+        self.cid.hash().to_bytes()
     }
 
-    /// Returns a base-58 encoded string of this `PeerId`.
+    /// Returns a raw bytes representation of the `PeerId`.
+    pub fn to_cidv1_bytes(&self) -> Vec<u8> {
+        self.cid.to_bytes()
+    }
+
+    /// Returns a base-58 raw encoded string of this `PeerId`.
     pub fn to_base58(&self) -> String {
         bs58::encode(self.to_bytes()).into_string()
```

Comment:
The following would fail given that `<PeerId as FromStr>` expects the string to start with `Qm` or `1`, right?

```rust
assert!(cidv1_peer_id == PeerId::from_str(cidv1_peer_id.to_base58()));
```

I am not quite sure how to proceed in this case. Maybe convert the CIDv1 peer ID to the old format first?

Reply:
> Maybe convert the CIDv1 peer ID to the old format first?

What do you think @JerryHue?